### PR TITLE
KAFKA-3786: Let ConfigDef filter property key value pairs

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by guozhangwang** [comment]
@ijuma This is just one proposal for tackling the problem I mentioned in JIRA. Another related issue resolved here is that Kafka Streams may prefer a different default value for some of the underlying configs.

Let me know what do you think, and if there is a better approach, I'd be OK to go with that as well.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 1]
```diff
@@ -221,7 +239,18 @@ public StreamsConfig(Map<?, ?> props) {
...
+
+        // enforce streams overrides if not specified by user
+        for (String keyName : CONSUMER_OVERRIDES.keySet()) {
+            if (!props.containsKey(keyName))
+                props.put(keyName, CONSUMER_OVERRIDES.get(keyName));
+        }
+
+        // disable auto commit ignoring any user overridden values,
+        // this is necessary for streams commit semantics
+        props.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Is ignoring the right answer? Would it be better to fail fast? If not, at least a warning should be logged.

**Comment by ijuma** [comment]
@guozhangwang I think the general idea is fine, but not sure about some of the details. Will comment on the PR.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 2]
```diff
@@ -432,6 +432,24 @@ public ConfigDef withClientSaslSupport() {
     }
 
     /**
+     * Filter configs that are not defined in this configuration definition.
+     *
+     * @param props The configs to be filtered.
+     * @return Filtered configs.
+     */
+    public Map<String, Object> filter(Map<String, Object> props) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: `filter` takes a predicate where here we are passing a value. I think I would expose the known config keys from `ConsumerConfig` and `ProducerConfig` instead. And then `StreamsConfig` can do whatever it wants with it.

**Comment by ijuma** [comment]
One more thing: have you considered adding a `logUnused` to the `KafkaStreams` constructor? It seems like it would do the right thing if placed after `KafkaConsumer` and `KafkaProducer` are instantiated.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 2 above)
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: I thought about that, and was a bit worried about this approach since we have currently not exposed the config keys (they are considered private).

Let me see if there is a better way of exposing it.

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: That's a good point. Note that we only need the config key names and that is safe to expose.

**Comment by guozhangwang** [comment]
I think we still cannot call `logUnused` at the end because the configs may be used later when tasks / state managers are created, and users may want to pass extra config values that will only be used in `ProcessorContext`, plus we construct the produce / consumer configs by copying the property value.

**Comment by ijuma** [comment]
OK, fair enough.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 1 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Ack.

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java [hunk 3]
```diff
@@ -53,7 +53,7 @@ public AbstractConfig(ConfigDef definition, Map<?, ?> originals, boolean doLog)
                 throw new ConfigException(key.toString(), originals.get(key), "Key must be a string.");
         this.originals = (Map<String, ?>) originals;
         this.values = definition.parse(this.originals);
-        this.used = Collections.synchronizedSet(new HashSet<String>());
+        this.used = Collections.synchronizedSet(new HashSet<>());
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Turns out I still need to specify the types for Java 7.

Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: Yeah, inference doesn't work in that case. Very lame! Are you saying that they improved that in Java 8?

**Comment by guozhangwang** [comment]
@ijuma Mind taking another look?

File: clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java (hunk 3 above)
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java]: I thought so, since the unit test passed locally before.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java [hunk 4]
```diff
@@ -376,6 +377,10 @@ public static Properties addDeserializerToConfig(Properties properties,
         super(CONFIG, props);
     }
 
+    public static Map<String, ConfigKey> configKeys() {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java]: Should this be `configNames` instead with an implementation like:

``` java
CONFIG.names()
```

It doesn't seem like we actually need `ConfigKey`.

File: clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java [hunk 5]
```diff
@@ -336,6 +337,10 @@ public static Properties addSerializerToConfig(Properties properties,
         super(CONFIG, props);
     }
 
+    public static Map<String, ConfigKey> configKeys() {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/producer/ProducerConfig.java]: Same as above.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 6]
```diff
@@ -212,6 +218,24 @@
                                         CommonClientConfigs.METRICS_NUM_SAMPLES_DOC);
     }
 
+    // this is the list of configs for underlying clients
+    // that streams prefer different default values
+    private static final Map<String, Object> PRODUCER_OVERRIDES;
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Nit: maybe we should call this `PRODUCER_DEFAULT_OVERRIDES` and the other one `CONSUMER_DEFAULT_OVERRIDES`.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 7]
```diff
@@ -221,7 +245,23 @@ public StreamsConfig(Map<?, ?> props) {
...
-        Map<String, Object> props = getBaseConsumerConfigs();
+        // filter out non-consumer configs
+        Map<String, ConfigKey> configKeys = ConsumerConfig.configKeys();
+        Map<String, Object> props = filter(configKeys, this.originals());
+
+        // disable auto commit ignoring any user overridden values,
+        // this is necessary for streams commit semantics
+        if (props.containsKey(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG)) {
+            log.warn("Ignoring user specified consumer config " + ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG
+                    + " and always turn off auto committing.");
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Can you please explain why it's better to warn than to fail-fast in this case? Just want to make sure I understand the reasoning for the choice.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 8]
```diff
@@ -232,15 +272,30 @@ public StreamsConfig(Map<?, ?> props) {
...
+            props.remove(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG);
+        }
+
+        // enforce streams overrides if not specified by user
+        for (String keyName : CONSUMER_OVERRIDES.keySet()) {
+            if (!props.containsKey(keyName))
+                props.put(keyName, CONSUMER_OVERRIDES.get(keyName));
+        }
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: It seems like the code here is the same as the code in `getConsumerConfigs`. Probably worth extracting into a private method. Why did you remove `getBaseConsumerConfigs`?

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java [hunk 9]
```diff
@@ -52,22 +54,26 @@ public void setUp() {
...
+        assertEquals(returnedProps.get(ConsumerConfig.MAX_POLL_RECORDS_CONFIG), "1000");
+        assertNull(returnedProps.get("DUMMY"));
     }
 
     @Test
     public void testGetRestoreConsumerConfigs() throws Exception {
         Map<String, Object> returnedProps = streamsConfig.getRestoreConsumerConfigs("client");
         assertEquals(returnedProps.get(ConsumerConfig.CLIENT_ID_CONFIG), "client-restore-consumer");
         assertNull(returnedProps.get(ConsumerConfig.GROUP_ID_CONFIG));
+        assertNull(returnedProps.get("DUMMY"));
```
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:76]: Should probably check that the consumer and producer configs no longer have properties that are not defined in their config.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 8 above)
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: In fact, even the producer code is pretty similar, so you could have a methods like the following which is called by both `getBaseConsumerConfigs` and `getProducerConfigs`:

``` java
private Map<String, Object> clientProps(Set<String> configNames, Map<String, Object> originals, Map<String, Object> overrides) {    
    Map<String, Object> props = filter(configNames, originals);
    // enforce streams overrides if not specified by user
    for (String keyName : overrides.keySet()) {
        if (!props.containsKey(keyName))
            props.put(keyName, overrides.get(keyName));
    }
}
```

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 10]
```diff
@@ -212,6 +218,24 @@
...
+    // this is the list of configs for underlying clients
+    // that streams prefer different default values
+    private static final Map<String, Object> PRODUCER_OVERRIDES;
+    static
+    {
+        PRODUCER_OVERRIDES = new HashMap<>();
+        PRODUCER_OVERRIDES.put(ProducerConfig.LINGER_MS_CONFIG, "100");
+    }
+
+    private static final Map<String, Object> CONSUMER_OVERRIDES;
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: For cases such as this, I think it's good practice to have a temp variable where the map is populated and then wrap it in a Collections.unmodifiableMap while assigning it to the static field. That way we're sure the map is not modified accidentally.

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java (hunk 4 above)
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java]: Ack on both.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 6 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Ack.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 8 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Ack. Combining this with filter as well.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 7 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: I do not have a good reason, other than "throwing one exception per thread may be worse than having one warn entry per thread", but maybe in the future we can have the `KafkaStreams` client to capture these exceptions and only throw them as one. So converting to fail-fast now.

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java (hunk 9 above)
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:76]: I thought we did with the `DUMMY` config?

Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:76]: Ack.

**Comment by guozhangwang** [comment]
@ijuma Addressed comments.

File: streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java (hunk 9 above)
Comment (ijuma) [inline streams/src/test/java/org/apache/kafka/streams/StreamsConfigTest.java:76]: Sorry, my bad, you are right.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 11]
```diff
@@ -212,6 +215,28 @@
...
 
+    // this is the list of configs for underlying clients
+    // that streams prefer different default values
+    private static final Map<String, Object> PRODUCER_DEFAULT_OVERRIDES;
+    static
+    {
+        Map<String, Object> tempProducerDefaultOverrides = new HashMap<>();
+        tempProducerDefaultOverrides.put(ProducerConfig.LINGER_MS_CONFIG, "100");
+
+        PRODUCER_DEFAULT_OVERRIDES = new HashMap<>(tempProducerDefaultOverrides);
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: I was thinking that this should be `Collections.unmodifiableMap(tempProducerDefaultOverrides)` to ensure that it is never mutated after it's assigned.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 12]
```diff
@@ -212,6 +215,28 @@
...
+
+    private static final Map<String, Object> CONSUMER_DEFAULT_OVERRIDES;
+    static
+    {
+        Map<String, Object> tempConsumerDefaultOverrides = new HashMap<>();
+        tempConsumerDefaultOverrides.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, "1000");
+        tempConsumerDefaultOverrides.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
+        tempConsumerDefaultOverrides.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false");
+
+        CONSUMER_DEFAULT_OVERRIDES = new HashMap<>(tempConsumerDefaultOverrides);
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: Same as above.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 13]
```diff
@@ -220,8 +245,18 @@ public StreamsConfig(Map<?, ?> props) {
         super(CONFIG, props);
     }
 
-    public Map<String, Object> getConsumerConfigs(StreamThread streamThread, String groupId, String clientId) {
-        Map<String, Object> props = getBaseConsumerConfigs();
+    public Map<String, Object> getConsumerConfigs(StreamThread streamThread, String groupId, String clientId) throws ConfigException {
+        // first disable auto commit ignoring any user overridden values,
+        // this is necessary for streams commit semantics
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java]: This comment needs to be updated, we now fail instead of disabling and ignoring.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 14]
```diff
@@ -232,15 +267,24 @@ public StreamsConfig(Map<?, ?> props) {
...
 
         return props;
     }
 
-    public Map<String, Object> getRestoreConsumerConfigs(String clientId) {
-        Map<String, Object> props = getBaseConsumerConfigs();
+    public Map<String, Object> getRestoreConsumerConfigs(String clientId) throws ConfigException {
+        // first disable auto commit ignoring any user overridden values,
+        // this is necessary for streams commit semantics
+        Map<String, Object> originals = this.originals();
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:278]: The comment needs to be updated as in the other method.

**Comment by ijuma** [comment]
Left a few minor comment, the rest LGTM. Feel free to merge after fixing them.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 15]
```diff
@@ -232,15 +267,24 @@ public StreamsConfig(Map<?, ?> props) {
...
+
+        if (originals.containsKey(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG)) {
+            throw new ConfigException("Unexpected user-specified consumer config " + ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG
+                    + ", as the streams client will always turn off auto committing.");
+        }
+
+        // generate consumer configs from original properties and overridden maps
+        Map<String, Object> props = clientProps(ConsumerConfig.configNames(), originals, CONSUMER_DEFAULT_OVERRIDES);
```
Comment (ijuma) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:288]: I probably would have these 12 lines in a `consumerProps` method and also reuse it from `getConsumerConfigs`. But if you think it's better this way, then that's OK.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 14 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:278]: Ack on all.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 15 above)
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:288]: Since the disabling of auto commit is specific to Streams, I'd prefer to keep them here.

**Comment by guozhangwang** [comment]
Thanks @ijuma , addressed comments and merged to trunk.
